[RDY] Create commands using function composition

gfokkema opened the pull request:
This removes the need to create 'child commands' in for example the TransformCommand.

gfokkema committed:
TransformCommand: remove subcommands

Every subcommand is now promoted to a "real" command, e.g. what used to be
`:transform run <expression>` is now `:run <expression>`. This is in preparation
of the command refactor that follows in the next commits.

This also removes argument parsing and visitor functionality from the
command class.

gfokkema committed:
Compose commands from several chained functional interfaces.

gfokkema committed:
Remove all obsolete commands.

gfokkema committed:
Make ATransform and PTransform consistent with each other.

Unia commented:
Perhaps cherry-pick 772f47061b377e2dc9563b5d3599e75f1acc5efa, 88357eceffe624766b3001f32312b8b55610fec0, 5b8801939ff0ac7504a2de4e43483cb2174ccda2, 4fa1a5c3f286ede7ea58972043c6bee611070948 and 97cb63e7c77815828b988326dc1a724e5c4efaa6 from the [refactor-commands](https://github.com/spoofax-shell/spoofax-shell/commits/refactor-commands) branch?

gfokkema committed:
Rename execute to apply, according to the Function<A,R> interface.

gfokkema committed:
Create a somewhat real builder pattern.

gfokkema committed:
Fix exceptions and returning of invalid ISpoofaxResults.

This is achieved by wrapping invalid results in an empty unit.

gfokkema committed:
Use java 8 this:: sugar for passing lambda's.

gfokkema committed:
Rewrite EvaluateCommand as [PA]EvalFunction

Balletie committed:
Use visitor pattern and monadic composition for returning errors.

This has the following changes:

- IHook has been renamed to IResult, and IDisplay to IResultVisitor.
  The IResultVisitor, together with IResult and its subclasses,
  implement the visitor pattern.
- Introduce FailOrSuccessResult, which is our more domain-specific
  equivalent of an Either Monad. It has a method flatMap, which applies
  the given function only if the result represents a success, and
  otherwise returns the failed result. We use FailableFunction to
  compose functions that take an input and return a FailOrSuccessResult.
  It is our domain-specific equivalent of a Kleisli Arrow.
- Introduce FailResult and ExceptionResult. The former is for wrapping
  invalid ISpoofaxResults, the latter is created and returned when an
  exception occurs somewhere in the pipeline of execution.
- A small fix to ParseResult#styled so that it returns the messages when
  the ParseUnit is invalid, just like the other ISpoofaxResult classes
  do.

Balletie committed:
Extract lambda to InputFunction, and create factory method for it

This removes the resultFactory as dependency from the CommandBuilder

Balletie committed:
Retry parsing with default start symbol when the first try was invalid

gfokkema committed:
Fix ParseCommandTest using the new monad functions.

gfokkema committed:
Changed method signature to allow passing lambda's.

This is used in the junit tests to stub parse functions.

gfokkema committed:
Fix guice loading errors introduced by previous commit.

gfokkema committed:
Fix junit for analyzecommand. OrElse is executed eagerly.

gfokkema committed:
Move tests and add more of them.

gfokkema committed:
Fix javadoc, build should (almost) pass again.

Balletie committed:
Create method for retrieving the source text of every SpoofaxResult

For this the TransformResult needed to be split up in two classes, to get the source text of wrapped analyzed/parsed units

Balletie committed:
Check if the ParseUnit is successful in valid, to never return a null AST

Balletie committed:
Test all AbstractSpoofaxResult subclasses with one parameterized test

gfokkema committed:
Create junit tests for transform function.

Balletie committed:
Replace occurences of IDisplay with IResultVisitor, IHook with IResult

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/impl/ConsoleRepl.java:
```diff
@@ -31,12 +31,12 @@
      */
     @Inject
-    public ConsoleRepl(IEditor editor, IDisplay display, ICommandInvoker invoker) {
+    public ConsoleRepl(IEditor editor, IResultVisitor display, ICommandInvoker invoker) {
```
Can you change the parameter and instance variable name as well?

Hjdskes commented on org.metaborg.spoofax.shell.console/src/test/java/org/metaborg/spoofax/shell/client/console/impl/MockModule.java:
```diff
@@ -12,11 +12,11 @@
     private final ICommandInvoker invokerMock;
     private final IEditor editorMock;
-    private final IDisplay displayMock;
+    private final IResultVisitor displayMock;
```
Same as above: can you rename this variable?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/client/IRepl.java:
```diff
@@ -22,13 +22,15 @@
      * @param input
      *            The input to send for evaluation.
-     * @return An {@link IHook} to process the result of the evaluation.
+     * @return An {@link IResult} to process the result of the evaluation, or {@code null} when an
```
Merge-gone-wrong: this JavaDoc is outdated: an "empty" `IResult` is returned when the input is empty.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/client/IRepl.java:
```diff
@@ -22,13 +22,15 @@
      *             When the command could not be found.
      */
-    default IHook eval(String input) throws MetaborgException, CommandNotFoundException {
+    // TODO: return Optional?
```
Merge-gone-wrong. This should be removed.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/client/IResult.java:
```diff
@@ -9,9 +9,9 @@
  * that hooks are returned is to allow the client to process it wherever and whenever it needs.
```
Please update the JavaDoc accordingly

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/client/IResultVisitor.java:
```diff
@@ -0,0 +1,43 @@
+     * Visit an error.
+     *
+     * @param errorResult
+     *            The {@link IResult} representing an error.
```
`The {@link FailResult} representing an error.`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/CommandBuilder.java:
```diff
@@ -0,0 +1,202 @@
+import com.google.inject.assistedinject.AssistedInject;
+
+/**
+ * Builds commands by composing several smaller functional interfaces.
```
Which are? You can (and should, imo) put more effort in the JavaDoc. Our work is going to be picked up by others, so having good documentation is a must.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/CommandBuilder.java:
```diff
@@ -0,0 +1,202 @@
+        return new IReplCommand() {
+            @Override
+            public IResult execute(String... arg) throws MetaborgException {
+                return function.apply(arg[0]);
```
This limits the functionality of all commands to only one string argument. With extensibility in mind, is that going to be a problem?

Balletie replied:
Good point.

Balletie replied:
Has been fixed.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/HelpCommand.java:
```diff
@@ -67,7 +67,7 @@ public IHook execute(String... args) throws MetaborgException {
-                .displayMessage(new StyledText(formathelp(commands)));
+                .visitMessage(new StyledText(formathelp(commands)));
         } catch (CommandNotFoundException e) {
             throw new MetaborgException("Command not found: " + e.commandName());
```
Should this be a `visitThrown`, or is this caught and handled elsewhere that I haven't seen yet?

Balletie replied:
This would be caught by the client I think.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/LanguageCommand.java:
```diff
@@ -85,27 +98,35 @@ private FileObject resolveLanguage(String path) {
         }
-
-        invoker.addCommand("eval", commandFactory.createEvaluate(project, lang, analyze));
+        new TransformVisitor(menuService).getActions(lang).forEach((key, action) -> {
```
Can't you inject the menu service in the `TransformVisitor`? It saves you a constructor parameter and is more correct, I think.

Balletie replied:
It's fine

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/TransformVisitor.java:
```diff
@@ -0,0 +1,67 @@
+import com.google.inject.Inject;
+
+/**
+ * An {@IMenuItemVisitor} implementation to create one {@link TransformCommand} per menu item.
```
`{@link IMenuItemVisitor}`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/TransformVisitor.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    /**
+     * Retrieve all the actions from the {@link IMenuService} belonong to the {@link ILanguageImpl}.
```
`belonging`.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AEvalFunction.java:
```diff
@@ -0,0 +1,59 @@
+ * Creates an {@link EvaluateResult} from a given {@link AnalyzeResult}.
+ */
+public class AEvalFunction extends AbstractSpoofaxFunction<AnalyzeResult, EvaluateResult> {
+    private IContextService contextService;
```
Can be `final`.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AEvalFunction.java:
```diff
@@ -0,0 +1,59 @@
+    private IContextService contextService;
+
+    @Inject
+    private Map<String, IEvaluationStrategy> evaluationStrategies;
```
Why inject this here instead of via the constructor?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AEvalFunction.java:
```diff
@@ -0,0 +1,59 @@
+    private Map<String, IEvaluationStrategy> evaluationStrategies;
+
+    /**
+     * Instantiate an {@link PEvalFunction}.
```
`Instantiate an {@link AEvalFunction}.`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/ATransformFunction.java:
```diff
@@ -0,0 +1,73 @@
+    private final ITransformAction action;
+
+    /**
+     * Instantiate a new {@link AnalyzedTransformCommand}.
```
`Instantiate a new {@link ATransformFunction}.`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/ATransformFunction.java:
```diff
@@ -0,0 +1,73 @@
+            context = a.context().get();
+        } else {
+            context = contextService.get(a.source(), project, lang);
+        }
```
Why not `IContext context = a.context().orElse(contextService.get(a.source(), project, lang));`?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AbstractSpoofaxFunction.java:
```diff
@@ -0,0 +1,67 @@
+    protected final ILanguageImpl lang;
+
+    /**
+     * Instantiate a {@link AbstractSpoofaxCommand}.
```
`Instantiate an {@link AbstractSpoofaxFunction}.`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AnalyzeFunction.java:
```diff
@@ -0,0 +1,67 @@
+import com.google.inject.assistedinject.Assisted;
+
+/**
+ * Represents a parse command sent to Spoofax.
```
`an analyze command`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AnalyzeFunction.java:
```diff
@@ -0,0 +1,67 @@
+    private final ISpoofaxAnalysisService analysisService;
+
+    /**
+     * Instantiate an {@link AnalyzeCommand}.
```
`{@link AnalyzeFunction}.`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AnalyzeFunction.java:
```diff
@@ -0,0 +1,67 @@
+            context = a.context().get();
+        } else {
+            context = contextService.get(a.source(), project, lang);
+        }
```
Why not `IContext context = a.context().orElse(contextService.get(a.source(), project, lang));`?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/FailableFunction.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    default <NewSuccess extends ISpoofaxResult<?>> FailableFunction<In, NewSuccess, Fail>
+        kleisliCompose(FailableFunction<Success, NewSuccess, Fail> other) {
+        Objects.requireNonNull(other);
```
Your JavaDoc should mention that `other` cannot be `null`, or you should use the `@NonNull` annotation here.

Balletie replied:
It would be annotated `@Nullable` if it could be null, so I will keep it this way.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/PEvalFunction.java:
```diff
@@ -0,0 +1,59 @@
+ * Creates an {@link EvaluateResult} from a given {@link ParseResult}.
+ */
+public class PEvalFunction extends AbstractSpoofaxFunction<ParseResult, EvaluateResult> {
+    private IContextService contextService;
```
Can be `final`.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/PEvalFunction.java:
```diff
@@ -0,0 +1,59 @@
+    private IContextService contextService;
+
+    @Inject
+    private Map<String, IEvaluationStrategy> evaluationStrategies;
```
Why is this injected here as opposed to the constructor?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/PTransformFunction.java:
```diff
@@ -0,0 +1,74 @@
+            context = a.context().get();
+        } else {
+            context = contextService.get(a.source(), project, lang);
+        }
```
Why not `IContext context = a.context().orElse(contextService.get(a.source(), project, lang));`?

Balletie replied:
Because java does eager evaluation, so we would make a context even if the `Optional` is not empty.

And `Optionall::orElseGet` (the "lazy" equivalent) would not work either because a `Supplier` may not throw an exception.

Hjdskes replied:
Then are these here correct?

https://github.com/spoofax-shell/spoofax-shell/pull/30/files#diff-c657d6b79bbb22e6d28451da0e0b70c9R52

https://github.com/spoofax-shell/spoofax-shell/pull/30/files#diff-a8a05e15fedf2f6b8151c706e1ed96aeR52

Balletie replied:
no, good catch.

Balletie replied:
I will try to remove this duplication.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/ParseFunction.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class ParseFunction extends AbstractSpoofaxFunction<InputResult, ParseResult> {
+    private final ISpoofaxSyntaxService syntaxService;
+    private ISpoofaxUnitService unitService;
```
Can be `final`.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/EvaluateResult.java:
```diff
@@ -7,14 +7,14 @@
 
 /**
- * The result of the execution of an {@link EvaluateCommand}.
+ * The result of the execution of an {@link PEvalFunction}.
```
And also an `AEvalFunction`?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/FailOrSuccessResult.java:
```diff
@@ -0,0 +1,144 @@
+     * an error.
+     *
+     * @param result
+     *            The cause of the error.
```
Incorrect JavaDoc.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/FailOrSuccessResult.java:
```diff
@@ -0,0 +1,144 @@
+     * turn be a success or an error.
+     *
+     * @param errorFunc
+     *            The {@link FailableFunction} to apply when this was a success.
```
So you execute `_error_Func` when this was a _success_?

Balletie replied:
it should probably be named failableFunc like the class

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/FailOrSuccessResult.java:
```diff
@@ -0,0 +1,144 @@
+        public <NewS extends ISpoofaxResult<?>> FailOrSuccessResult<NewS, F>
+            flatMap(FailableFunction<S, NewS, F> errorFunc) {
+            Objects.requireNonNull(errorFunc);
+            return errorFunc.apply(result);
```
`errorFunc` in a `Successful`?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/FailOrSuccessResult.java:
```diff
@@ -0,0 +1,144 @@
+     */
+    private static final class Failed<S extends IResult, F extends IResult>
+        extends FailOrSuccessResult<S, F> {
+        private F result;
```
Can be `final`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/FailOrSuccessResult.java:
```diff
@@ -0,0 +1,144 @@
+     */
+    private static final class Successful<S extends IResult, F extends IResult>
+        extends FailOrSuccessResult<S, F> {
+        private S result;
```
Can be `final`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/ParseResult.java:
```diff
@@ -50,12 +49,20 @@ public ParseResult(IStrategoCommon common,
+    @Override
     public boolean valid() {
-        return unit().valid();
+        return unit().valid() && unit().success();
```
Other result classes do not require this?

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/TransformResult.java:
```diff
@@ -7,28 +7,85 @@
-    @AssistedInject
-    public TransformResult(IStrategoCommon common,
-                           @Assisted ISpoofaxTransformUnit<?> unit) {
+    @Inject
```
`@AssistedInject`

Balletie replied:
Not necessary as there's just one ctor, but I'll change it anyway

Hjdskes commented on org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/output/SpoofaxResultsTest.java:
```diff
@@ -0,0 +1,277 @@
+import org.spoofax.interpreter.terms.IStrategoTerm;
+
+/**
+ * Tests all subclasses of {@link AbstractSpoofaxResult} using a {@link Parameterized parameterized
```
Sweet! :D

Hjdskes commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/EclipseRepl.java:
```diff
@@ -33,20 +33,20 @@
      *            The {@link ICommandInvoker} for executing user input.
-     * @param display
+     * @param visitor
      *            The {@link EclipseDisplay} to send results to.
```
Please update this as well.

Hjdskes commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/EclipseRepl.java:
```diff
@@ -104,11 +104,11 @@ protected IStatus run(IProgressMonitor monitor) {
 
-    private void runAsUIJob(IHook hook) {
+    private void runAsUIJob(IResult hook) {
         Job job = new UIJob("Spoofax REPL display job") {
```
And perhaps this one too.

Balletie committed:
Add abstract base class for Functions evaluated in an IContext

Unia commented:
Collapsed but still unresolved comments:

1. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66868783
2. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66869598
3. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66870297

Balletie committed:
Add EvaluateFunctionTest, inject evalStrategies via constructor for it.

Balletie committed:
Improve API of CommandBuilder by allowing custom composition

This makes the CommandBuilder more reusable and extensible. The
changes:

1. Let FailableFunction be composed by a FailableFunction returning an
IResult, instead of just ISpoofaxResult subclasses.
2. Add compose function to CommandBuilder to allow library users to
compose their own functions and build their own commands.
3. Add possibility for setting the initial function of a CommandBuilder.

Balletie committed:
Update javadoc and parameter names to IResultVisitor intro

Balletie committed:
Let CommandBuilder build commands accepting varargs.

This makes the builder more extensible

Balletie committed:
Add javadoc class comment to CommandBuilder

Balletie commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/IReplCommand.java:
```diff
@@ -20,9 +20,9 @@
      * @param args
      *            The arguments for this command.
-     * @return An {@link IHook} to process the result of this command.
+     * @return An {@link IResult} to process the result of this command.
```
Updated this one too.

Balletie committed:
Workaround for build failure introduced by two commits earlier

See https://bugs.eclipse.org/bugs/show_bug.cgi?id=436542

Balletie committed:
Process rest of feedback by @Unia

- Update/improve JavaDoc.
- final fields where possible.
- Better parameter name

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AEvalFunction.java:
```diff
@@ -0,0 +1,59 @@
+    private final Map<String, IEvaluationStrategy> evaluationStrategies;
+
+    /**
+     * Instantiate an {@link PEvalFunction}.
```
`AEvalFunction`

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/AnalyzeFunction.java:
```diff
@@ -0,0 +1,59 @@
+    private final ISpoofaxAnalysisService analysisService;
+
+    /**
+     * Instantiate an {@link AnalyzeCommand}.
```
`AnalyzeFunction`

Unia commented:
1. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66867797 -- do we keep it as is? I assume so.
2. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66867921
3. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66868327
4. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66869986
5. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66871675
6. https://github.com/spoofax-shell/spoofax-shell/pull/30#discussion-diff-66872114 -- you can rename `EclipseDisplay` to `EclipseVisitor`, maybe. Or we just keep `EclipseDisplay`.

Balletie committed:
Fix bugs found with testing and from feedback @Unia

1. Check success() and valid() for each SpoofaxResult.
2. Check if there's actually an AST when pretty printing, to avoid null pointers
